File: thisweek_masuipeo/tyama_codeiq1004.py
#Used this to solve checkio "Lantern Festival"

D=( #Right,Up,Left,Down
	(0,1),
	(-1,0),
	(0,-1),
	(1,0),
)

class Maze:
	def __init__(self,x,y,gx,gy,d,v):
		self.x=x
		self.y=y
		self.gx=gx
		self.gy=gy
		self.d=d
		self.v=v
	def ok(self):
		return self.gx==self.x and self.gy==self.y
	def same(self,other):
		return self.x==other.x and self.y==other.y
	def move(self):
		d_orig=(self.d+3)%4
		for i in range(4):
			self.d=(d_orig+i)%4
			if self.v[self.y+D[self.d][0]][self.x+D[self.d][1]]!='#': break
		self.y=self.y+D[self.d][0]
		self.x=self.x+D[self.d][1]
		#printf("%d %d %d\n"%(x,y,d))

def solvable(_x,_y,_gx,_gy,v):
	if v[_y][_x]=='#' or v[_gy][_gx]=='#': return False
	q=[]
	back={}
	q.append((_x,_y))
	back[(_x,_y)]=(_x,_y)
	while q:
		x,y=q[0]
		q.pop(0)
		for d in range(4):
			if v[y+D[d][0]][x+D[d][1]]!='#':
				_p=(x+D[d][1],y+D[d][0])
				if _p==(_gx,_gy): return True
				if _p not in back:
					back[_p]=(x,y)
					q.append(_p)
	return False

if __name__=='__main__':
	N=4
	power=N*N
	v=[]
	v.append(['#']*(N+2))
	for i in range(N):
		v.append(['#']+[' ']*N+['#'])
	v.append(['#']*(N+2))
	#v=[
	#	list("######"),
	#	list("#    #"),
	#	list("##   #"),
	#	list("# #  #"),
	#	list("#    #"),
	#	list("######"),
	#]
	r=0
	for i in range(1<<power):
		for j in range(power):
			v[j/N+1][j%N+1]='#' if (i&(1<<j)) else ' '
		if not solvable(1,1,N,N,v): continue
		m1=Maze(1,1,N,N,3,v)
		m2=Maze(N,N,1,1,1,v)
		while not m1.ok() and not m2.ok():
			m1.move()
			m2.move()
			if m1.same(m2):
				r+=1
				break
	print(r)

File: thisweek_masuipeo/tyama_codeiq1004.cpp
/*
手元で試してみましたが、増井さんが示してくださった、
	vector<string>v={
		"######",
		"#    #",
		"#  # #",
		"#   ##",
		"#    #",
		"######",
	};
については出会いませんでした
(Aさんがゴールした7手目で終わりになったからだと思われます)。
しかし、その点対称である
	vector<string>v={
		"######",
		"#    #",
		"##   #",
		"# #  #",
		"#  X #",
		"######",
	};
については上図のXで出会ってしまいました。
m1.ok()とm2.ok()の両方が偽である必要があるという点を直しました。
*/

#include <cstdio>
#include <string>
#include <queue>
#include <map>
using namespace std;
const int D[4][2]={ //Right,Up,Left,Down
	{0,1},
	{-1,0},
	{0,-1},
	{1,0},
};

class Maze{
	const vector<string> &v;
	const int gx,gy;
	int x,y,d;
public:
	Maze(int _x,int _y,int _gx,int _gy,int _d,const vector<string>&_v)
		: x(_x),y(_y),gx(_gx),gy(_gy),d(_d),v(_v){}
	bool ok(){return x==gx && y==gy;}
	bool same(Maze& other){return x==other.x && y==other.y;}
	void move(){
		int d_orig=(d+3)%4,i=0;
		for(;i<4;i++){
			d=(d_orig+i)%4;
			if(v[y+D[d][0]][x+D[d][1]]!='#')break;
		}
		y=y+D[d][0];
		x=x+D[d][1];
		//printf("%d %d %d\n",x,y,d);
	}
};

bool solvable(int _x,int _y,int _gx,int _gy,const vector<string>&v){
	if(v[_y][_x]=='#' || v[_gy][_gx]=='#')return false;
	typedef pair<int,int> pii;
	queue<pii>q;
	map<pii,pii>back;
	q.push(make_pair(_x,_y));
	back[make_pair(_x,_y)]=make_pair(_x,_y);
	for(;!q.empty();){
		pii p=q.front();q.pop();
		int x=p.first,y=p.second;
		for(int d=0;d<4;d++){
			if(v[y+D[d][0]][x+D[d][1]]!='#'){
				pii _p=make_pair(x+D[d][1],y+D[d][0]);
				if(_p==make_pair(_gx,_gy))return true;
				if(back.find(_p)==back.end())back[_p]=p,q.push(_p);
			}
		}
	}
	return false;
}

int main(){
	const int N=5;
	const int power=N*N;
	vector<string>v(N+2);
	v[0]=v[N+1]=string(N+2,'#');
	for(int i=1;i<=N;i++){
		v[i]=string(1,'#')+string(N,' ')+'#';
	}
	/*
	vector<string>v={
		"######",
		"#    #",
		"##   #",
		"# #  #",
		"#    #",
		"######",
	};
	*/
	int r=0;
	for(int i=0;i<(1<<power);i++){
		for(int j=0;j<power;j++)v[j/N+1][j%N+1]=(i&(1<<j)) ? '#' : ' ';
		if(!solvable(1,1,N,N,v))continue;
		Maze m1(1,1,N,N,3,v);
		Maze m2(N,N,1,1,1,v);
		for(;!m1.ok()&&!m2.ok();){
			m1.move();
			m2.move();
			if(m1.same(m2)){
				r++;
				//for(int k=0;k<N+2;k++)printf(v[k].c_str()),printf(",");puts("");
				//puts("OK!");
				break;
			}
		}
	}
	printf("%d\n",r); //660148
}